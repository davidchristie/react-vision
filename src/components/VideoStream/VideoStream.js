import PropTypes from 'prop-types'
import React, { Component } from 'react'

export default class extends Component {
  static propTypes = {
    stream: PropTypes.object
  }

  componentWillReceiveProps (nextProps) {
    const { stream } = nextProps
    if (stream && this.video) {
      this.video.srcObject = stream
    }
  }

  getImage () {
    if (this.video) {
      const canvas = document.createElement('canvas')
      canvas.width = this.video.videoWidth
      canvas.height = this.video.videoHeight
      canvas.getContext('2d')
        .drawImage(this.video, 0, 0, canvas.width, canvas.height)
      return canvas.toDataURL()
    } else {
      return null
    }
  }

  render () {
    const { stream } = this.props
    return (
      <video
        autoPlay
        className='VideoStream'
        ref={element => {
          if (element) {
            this.video = element
            if (stream) {
              element.srcObject = stream
            }
          }
        }}
      />
    )
  }
}
