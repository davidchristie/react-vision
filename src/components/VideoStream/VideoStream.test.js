import { shallow, mount } from 'enzyme'
import React from 'react'

import VideoStream from './VideoStream'

it('mounts without stream', () => {
  expect(mount(<VideoStream />)).toMatchSnapshot()
})

it('mounts with stream', () => {
  const props = {
    stream: {}
  }
  expect(mount(<VideoStream {...props} />)).toMatchSnapshot()
})

describe('when the stream is changed', () => {
  it('updates the video srcObject if mounted', () => {
    const stream = {}
    const wrapper = mount(<VideoStream />)
    wrapper.setProps({
      stream
    })
    expect(wrapper.find('video').at(0).instance().srcObject).toEqual(stream)
  })

  it('runs without crashing if unmounted', () => {
    const stream = {}
    const wrapper = shallow(<VideoStream />)
    wrapper.setProps({
      stream
    })
  })
})

describe('getImage method', () => {
  beforeEach(() => {
    window.HTMLCanvasElement.prototype.getContext = () => ({
      drawImage: () => {}
    })
  })

  it('takes snapshot image', () => {
    const props = {
      stream: {}
    }
    const wrapper = mount(<VideoStream {...props} />)
    expect(wrapper.instance().getImage()).toBeDefined()
  })

  it('returns null if component is unmounted', () => {
    const wrapper = shallow(<VideoStream />)
    expect(wrapper.instance().getImage()).toBeNull()
  })
})
