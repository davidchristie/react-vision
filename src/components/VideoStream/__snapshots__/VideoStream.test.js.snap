// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mounts with stream 1`] = `
<_class
  stream={Object {}}
>
  <video
    autoPlay={true}
    className="VideoStream"
  />
</_class>
`;

exports[`mounts without stream 1`] = `
<_class>
  <video
    autoPlay={true}
    className="VideoStream"
  />
</_class>
`;
